lib/components/cluster.js

```diff
@@ -3,11 +3,116 @@
 import React from 'react';
 
 /**
+ * Types.
+ */
+
+type Props = {
+  children: Array<ReactElement>,
+  className: ?string,
+  height: number,
+  rowHeight: number,
+};
+
+type State = {
+  index: number,
+};
+
+/**
+ * Defaults and options.
+ */
+
+const INITIAL_STATE =  {
+  index: 0,
+};
+
+const RANGE_OFFSET = 5;
+
+/**
  * Cluster component.
+ *
+ * Displays large sets of data in a scroll container.
  */
 
-export default function Cluster(): ReactElement {
-  return (
-    <span>hello world</span>
-  );
+export default class Cluster extends React.Component {
+  props: Props;
+  state: State = INITIAL_STATE;
+
+  shouldComponentUpdate(nextProps: Props, nextState: State): boolean {
+    return this.props.children !== nextProps.children || this.state.index !== nextState.index;
+  }
+
+  componentDidMount(): void {
+    this.cluster.addEventListener('scroll', this.setIndex.bind(this));
+  }
+
+  componentWillUnmount(): void {
+    this.cluster.removeEventListener('scroll', this.setIndex.bind(this));
+  }
+
+  setIndex(): void {
+    this.setState({ index: this.findIndex() });
+  }
+
+  render(): ReactElement {
+    const {children, className, height, rowHeight} = this.props;
+
+    const rows = toClusterRows({
+      children,
+      height: rowHeight,
+      range: this.findRange(),
+    });
```

Comment:
this feels overly complex

eg 
- why does range pass findRange (vs having it in the function directly)
- why is this not simply a function that's called (eg renderRows)